src/gfx/qt.cpp

```diff
@@ -586,10 +614,175 @@ const char *GfxProcQT::supportedformatsQT()
         formatstring->resize(formatstring->size() - 1);
         formatstring->append(supportedformatsFfmpeg());
 #endif
+
+#ifdef HAVE_LIBRAW
+        formatstring->resize(formatstring->size() - 1);
+        formatstring->append(supportedformatsLibraw());
+#endif
     }
     return formatstring->constData();
 }
 
+#ifdef HAVE_LIBRAW
+
+const char *GfxProcQT::supportedformatsLibraw()
+{
+    return ".3fr.arw.cr2.crw.ciff.cs1.dcr.dng.erf.iiq.k25.kdc.mef.mos.mrw.nef.nrw.orf.pef.raf.raw.rw2.rwl.sr2.srf.srw.x3f.";
+}
+
+QImageReader *GfxProcQT::readbitmapLibraw(int &w, int &h, int &orientation, QString imagePath)
+{
+    LibRaw libRaw;
+    int ret = libRaw.open_file(imagePath.toUtf8().constData());
+    if (ret > 0 || LIBRAW_FATAL_ERROR(ret)
+            || libRaw.imgdata.sizes.width <= 0
+            || libRaw.imgdata.sizes.height <= 0)
+    {
+        LOG_debug << "Unreadable RAW image";
+        return NULL;
+    }
+
+    libraw_processed_image_t *output = NULL;
+    const libraw_data_t &imgdata = libRaw.imgdata;
+    LOG_debug << "Processing RAW image: " << imagePath.toUtf8().constData()
+              << " " << imgdata.sizes.width
+              << " " << imgdata.sizes.height
+              << " " << imgdata.thumbnail.twidth
+              << " " << imgdata.thumbnail.theight
+              << " " << imgdata.sizes.flip;
+
+    if (imgdata.thumbnail.twidth > 0 && imgdata.thumbnail.theight > 0)
+    {
+        ret = libRaw.unpack_thumb();
+        if (ret == 0 || (ret < 0 && !LIBRAW_FATAL_ERROR(ret)))
+        {
+            LOG_debug << "Extracting thumbnail from RAW image";
+            output = libRaw.dcraw_make_mem_thumb();
+        }
+    }
+
+    if (!output)
+    {
+        ret = libRaw.unpack();
+        if (ret == 0 || (ret < 0 && !LIBRAW_FATAL_ERROR(ret)))
+        {
+            LOG_debug << "Extracting full RAW image";
+            libRaw.dcraw_process();
+            output = libRaw.dcraw_make_mem_image();
+        }
+    }
+
+    if (!output)
+    {
+        LOG_warn << "Unable to extract RAW image";
+        return NULL;
+    }
+
+    QImage unscaled;
+    if (output->type == LIBRAW_IMAGE_JPEG)
+    {
+        LOG_debug << "Converting RAW image in JPG format";
+
+        unscaled.loadFromData(output->data, output->data_size, "JPEG");
+        w = unscaled.width();
+        h = unscaled.height();
+    }
+    else if (output->type == LIBRAW_IMAGE_BITMAP)
+    {
+        LOG_debug << "Converting RAW image in BITMAP format";
+
+        int numPixels = output->width * output->height;
+        int colorSize = output->bits / 8;
+        int pixelSize = output->colors * colorSize;
+
+        unscaled = QImage(output->width, output->height, QImage::Format_RGB32);
+        uchar *pixels = unscaled.bits();
+        uchar *data = output->data;
+        for (int i = 0; i < numPixels; i++, data += pixelSize)
+        {
+            if (output->colors == 3)
+            {
+                pixels[i * 4] = data[2 * colorSize];
```

Comment:
minor performance enhancement (in case the compiler don't do it automatically): save i*4

Reply:
I will change it 👍